Dart/thirdPartySrc/analysisServer/com/google/dart/server/internal/remote/processor/NotificationCompletionExistingImportsProcessor.java

```diff
@@ -0,0 +1,52 @@
+// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package com.google.dart.server.internal.remote.processor;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.dart.server.AnalysisServerListener;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+
+import java.util.*;
+
+public class NotificationCompletionExistingImportsProcessor extends NotificationProcessor {
+  public NotificationCompletionExistingImportsProcessor(AnalysisServerListener listener) { super(listener); }
+
+  /**
+   * Process the given {@link JsonObject} notification and notify {@link #listener}.
+   */
+  @Override
+  public void process(JsonObject response) throws Exception {
+    JsonObject paramsObject = response.get("params").getAsJsonObject();
+    String file = paramsObject.get("file").getAsString();
+    JsonObject existingImports = paramsObject.get("imports").getAsJsonObject();  // ExistingImports
+    Map<String, Set<String>> uriToNames = buildUriNamesMap(existingImports);
+    getListener().computedExistingImports(file, uriToNames);
+  }
+
+  private Map<String, Set<String>> buildUriNamesMap(JsonObject existingImports) {
```

Comment:
I think this is wrong.

This method should return something like `Map<importedLibraryUri, Map<declaringLibraryUri, declaredName>>`. What I see as a single `Map<declaringLibraryUri, declaredName>`, so we know *what* is imported into the target library, but not *how*. This is important for re-exports. For example both "material.dart" and "cupertino.dart" re-export `Widget`, and we need to know which suggestion set to use for `Widget`.

Reply:
Thanks @scheglov! I got confused and missed that one library can re-export multiple names.